File: 110_BalancedBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
//78.07%
class Solution {
public:
	bool isBalancedHelper(TreeNode *tree, int &height){
		height = 0;
		int heightL(0), heightR(0);
		if(!tree || isBalancedHelper(tree -> left, heightL) && isBalancedHelper(tree -> right, heightR) && abs(heightL - heightR) <= 1){
			height = heightL > heightR ? heightL + 1 : heightR + 1;
			return true;
		}
		return false;
	}
				
    bool isBalanced(TreeNode* root) {
        int heightL(0), heightR(0);
		if(!root || isBalancedHelper(root -> left, heightL) && isBalancedHelper(root -> right, heightR) && abs(heightL - heightR) <= 1)
			return true;
		return false;
    }
};

File: 110_BalancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# 78.37%
class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def heightBalanced(root):
            if not root:
                return 0, True
            heightL, flagL = heightBalanced(root.left)
            heightR, flagR = heightBalanced(root.right)
            if flagL and flagR and abs(heightL - heightR) <= 1:
                return max(heightL, heightR) + 1, True
            return 0, False
        
        if not root:
            return True
        heightL, flagL = heightBalanced(root.left)
        heightR, flagR = heightBalanced(root.right)
        if flagL and flagR and abs(heightL - heightR) <= 1:
            return True    
        return False